packages/store/addon/-private/system/weak-cache.ts

```diff
@@ -0,0 +1,73 @@
+import { DEBUG } from '@glimmer/env';
+
+import { addSymbol, symbol } from '../ts-interfaces/utils/symbol';
+
+interface WeakCache<K extends object, V> {
+  _symbol?: Symbol | string;
+  _fieldName?: string;
+  _generator?: (key: K) => V;
+  _expectMsg?: (key: K) => string;
+}
+
+class WeakCache<K extends object, V> {
+  private _cache = new WeakMap<K, V>();
+
+  constructor(_fieldName?: string, _generator?: (key: K) => V, _expectMsg?: (key: K) => string) {
+    this._generator = _generator;
```

Comment:
Another option for a name could be `defaultValueGenerator` or `valueGenerator` as the argument name and/or property on this class.  Both don't seem particularly appealing though.